td-reproducible-tool/src/main.rs

```diff
@@ -0,0 +1,200 @@
+
+use argparse::{ArgumentParser, Store, StoreTrue};
+use regex::bytes::Regex;
+use std::convert::TryInto;
+use std::fs;
+use std::io::Read;
+use std::{env, fs::File, path::PathBuf};
+use zeroize::Zeroize;
+
+// https://docs.microsoft.com/en-us/windows/win32/debug/pe-format
+const PE_SIG_PTR_OFF: usize = 0x3C;
+const TIMEDATASTAMP_OFF_TO_PE_SIG: usize = 0x8;
+
+fn main() -> std::io::Result<()> {
+    // handle args
+    let mut workspace = "".to_string();
+    let mut binary_name = "".to_string();
+    let mut target = "".to_string();
+    let mut profile = "".to_string();
+    let mut verbose = false;
+
+    let mut cargo_home = "".to_string();
+    let mut rustup_home = "".to_string();
+
+    {
+        let mut ap = ArgumentParser::new();
+        ap.set_description("TD REPRODUCIBLE TOOL");
+        ap.refer(&mut workspace).add_option(
+            &["-w", "--workspace"],
+            Store,
+            "Where to find the target folder.",
+        );
+        ap.refer(&mut binary_name).add_option(
+            &["-n", "--name"],
+            Store,
+            "Name for the compiled binary.",
+        );
+        ap.refer(&mut target)
+            .add_option(&["-t", "--target"], Store, "The built target to find.");
+        ap.refer(&mut profile).add_option(
+            &["-p", "--profile"],
+            Store,
+            "The built profile to find.",
+        );
+        ap.refer(&mut cargo_home)
+            .add_option(&["-c", "--cargo_home"], Store, "The cargo home.");
+        ap.refer(&mut rustup_home)
+            .add_option(&["-r", "--rustup_home"], Store, "The rustup home.");
+        ap.refer(&mut verbose)
+            .add_option(&["-v", "--verbose"], StoreTrue, "Verbose output.");
+        ap.parse_args_or_exit();
+    }
+
+    if target == "" {
+        target = ".".to_string();
+    } else if target == "x86_64-unknown-uefi" {
+        if !binary_name.ends_with(".efi") {
+            binary_name = binary_name + ".efi";
+        }
+    } else if target == "x86_64-unknown-none" {
+        // nothing to check
+    }
+
+    // solve the path
+    let binary_path: PathBuf = [
+        if workspace == "" { "." } else { &workspace },
+        "target",
+        &target,
+        if profile == "" { "release" } else { &profile },
+        &binary_name,
+    ]
+    .iter()
+    .collect();
+
+    assert!(binary_path.exists());
+    println!("INFO: Found the compiled file: {:?}", binary_path);
+
+    // check out CARGO_HOME and RUSTUP_HOME
+    cargo_home = if cargo_home == "" {
+        if env::var("CARGO_HOME").is_err() {
+            panic!("Neither --cargo_home nor system environment for \"CARGO_HOME\" is found! ")
+        } else {
+            env::var("CARGO_HOME").unwrap()
+        }
+    } else {
+        cargo_home
+    };
+
+    rustup_home = if rustup_home == "" {
+        if env::var("RUSTUP_HOME").is_err() {
+            panic!("Neither --rustup_home nor system environment for \"RUSTUP_HOME\" is found! ")
```

Comment:
Can we make it optional?

Reply:
It is optional now, please check the latest code change.